src/ws-handler.ts

```diff
@@ -321,14 +321,19 @@ export class WsHandler {
                 return;
             }
 
-            let member = { user_id, user_info };
+            this.server.adapter.getChannelMembers(ws.app.id, channel, false).then(members => {
+                let member = { user_id, user_info };
+
+                // If the member already exists in the channel, don't reconnect him.
+                if (members.has(user_id as string)) {
+                    return;
```

Comment:
Has this an impact on reconnects due to connection loss on the client side?

I guess if the client loses its connection or switches between networks (e.g. mobile and wlan), the server will only disconnect the client upon not receiving one or more pings in time. Can the client still connect and join a presence channel in the mean time, if that is the case; or will it need to wait until the server drops the old connection?

Reply:
@Namoshek I want to find a good way to make sure that a user can connect to a presence channel only once, otherwise, when multiple tabs/instances with/of different connections happen:

- other members will receive a false negative that the user joined (although it was already present in the room) by getting the `member_added` event
- when one of the tabs/instances drops, the user will be disconnected from all tabs/instances since the `member_removed` event will be broadcasted, the other members receiving a false negative that the user left (although it is still connected, but in another tab), and the server will also falsely store that the user is not in the channel.

Pusher recommends using service workers, but a hard check on the server-side is also needed. Or perhaps I may let multiple users connect via different connections and let the developers implement their approach on multiple tabs/instances for the same user in presence channels.

https://support.pusher.com/hc/en-us/articles/360019256174-How-can-I-prevent-multiple-tabs-opening-new-Channels-connections-

Reply:
Alright, thanks for the explanation - totally makes sense to handle this on the server-side, if possible!